File: pt4/1041.robot-bounded-in-circle.go
package main

/*
1041. Robot Bounded In Circle
*/
func isRobotBounded(instructions string) bool {
	x, y := 0, 0
	direction := 0
	moves := [][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	for i := 0; i < len(instructions); i++ {
		switch instructions[i] {
		case 'G':
			x += moves[direction][0]
			y += moves[direction][1]
		case 'L':
			direction = (direction + 3) % 4
		case 'R':
			direction = (direction + 1) % 4
		}
	}

	return (x == 0 && y == 0) || direction != 0
}

/*
Interview Explanation

Core idea:
Simulate one instruction cycle. The robot is bounded if it returns to the
origin, or if it finishes facing a direction other than north. A changed
direction rotates repeated displacements and closes within at most four cycles.

Go data structures:
- [][2]int maps direction index to movement vector.
- direction is an integer modulo 4: north, east, south, west.

Algorithm:
1. Start at (0,0) facing north.
2. Execute the instruction string once.
3. Return true if the robot is back at origin or no longer facing north.

Correctness:
If one cycle returns to origin, repetition stays bounded. If one cycle changes
direction, future cycles apply the same displacement rotated, and the net
displacement cancels after at most four cycles. If the robot is not at origin
and still faces north, each repetition adds the same displacement forever, so
it is unbounded.

Complexity:
Time is O(n), and space is O(1).

Edge cases:
- Only turns are bounded.
- "GG" is unbounded.
- "GL" is bounded despite not returning after one cycle.
*/

File: pt4/1041.robot-bounded-in-circle.java
/*
 * 1041. Robot Bounded In Circle
 */
class Solution {
    public boolean isRobotBounded(String instructions) {
        int x = 0;
        int y = 0;
        int direction = 0;
        int[][] moves = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

        for (int i = 0; i < instructions.length(); i++) {
            char instruction = instructions.charAt(i);
            if (instruction == 'G') {
                x += moves[direction][0];
                y += moves[direction][1];
            } else if (instruction == 'L') {
                direction = (direction + 3) % 4;
            } else {
                direction = (direction + 1) % 4;
            }
        }

        return (x == 0 && y == 0) || direction != 0;
    }
}

/*
Interview Explanation

Core idea:
Simulate one instruction cycle. The robot is bounded if it returns to the
origin, or if it ends facing a direction other than north. A changed direction
rotates the displacement on repeated cycles and closes within at most four
cycles.

Java data structures:
- int[][] moves maps direction index to (dx, dy).
- direction is an integer modulo 4: north, east, south, west.

Algorithm:
1. Start at (0, 0), facing north.
2. Simulate each character once.
3. Return true if position is origin or direction changed.

Correctness:
If one cycle returns to the origin, repeated cycles stay bounded. If one cycle
does not return but changes direction, future cycles apply the same movement
rotated, so the net displacement cancels after four cycles. If position is not
origin and direction is still north, each cycle adds the same displacement,
making the path unbounded.

Complexity:
Time is O(n), space is O(1).

Edge cases:
- Only turns are bounded.
- "GG" is unbounded.
- "GL" is bounded even though it is not back at origin after one cycle.
*/
